src/software/new_geom/cubic_bezier_spline2d_test.cpp

```diff
@@ -0,0 +1,247 @@
+
+    ASSERT_EQ(4, knot_vector.size());
+
+    for (size_t i = 0; i < knot_vector.size(); i++)
+    {
+        EXPECT_EQ(test_spline_1.getValueAt(knot_vector[i]), expected_knots[i]);
+    }
+    EXPECT_EQ(expected_knots, test_spline_1.getKnots());
+}
+
+TEST_F(CubicBezierSplineTest, getValueAt__start_point)
+{
+    // Check that the start point is at the correct position
+    const Point start_point = test_spline_1.getValueAt(0);
+    EXPECT_TRUE(TestUtil::equalWithinTolerance(start_point, Point(1, -1), 1e-9));
+
+    // Check that the tangent at the start is (approximately) what we expected. This
+    // serves to check that we're getting the value for intermediate points correctly
+    // as well.
+    const Point just_after_start_point   = test_spline_1.getValueAt(1e-9);
+    const Vector approx_tangent_at_start = just_after_start_point - start_point;
+    const Angle tangent_error_angle = approx_tangent_at_start.angleWith(Vector(-3, -4));
+
+    EXPECT_NEAR(0, tangent_error_angle.toRadians(), 1e-9);
+}
+
+TEST_F(CubicBezierSplineTest, getValueAt__end_point)
+{
+    // Check that the end point is at the correct position
+    const Point end_point = test_spline_1.getValueAt(1);
+    EXPECT_TRUE(TestUtil::equalWithinTolerance(end_point, Point(1, -2), 1e-9));
+
+    // Check that the tangent at the end is (approximately) what we expected. This
+    // serves to check that we're getting the value for intermediate points correctly
+    // as well.
+    const Point just_before_end_point  = test_spline_1.getValueAt(1 - 1e-9);
+    const Vector approx_tangent_at_end = just_before_end_point - end_point;
+    const Angle tangent_error_angle    = approx_tangent_at_end.angleWith(Vector(-2, -5));
+
+    EXPECT_NEAR(0, tangent_error_angle.toRadians(), 1e-9);
+}
+
+TEST_F(CubicBezierSplineTest, getStartPoint)
+{
+    EXPECT_TRUE(TestUtil::equalWithinTolerance(test_spline_1.getStartPoint(),
+                                               Point(1, -1), 1e-9));
+}
+
+TEST_F(CubicBezierSplineTest, getEndPoint)
+{
+    EXPECT_TRUE(
+        TestUtil::equalWithinTolerance(test_spline_1.getEndPoint(), Point(1, -2), 1e-9));
+}
+
+// Get the value at the knots (probably looking at individual segments and checking
+// start and end point)
+TEST_F(CubicBezierSplineTest, getKnots)
+{
+    std::vector<Point> expected_knots = {Point(1, -1), Point(5, 5), Point(10, -10),
+                                         Point(1, -2)};
+    EXPECT_EQ(expected_knots, test_spline_1.getKnots());
+}
+
+// Check that the spline is C_2 continuous at knots (get individual segments and check)
+TEST_F(CubicBezierSplineTest, check_c2_continuous_at_knots)
+{
+    std::vector<double> parametrization_values =
+        test_spline_1.getKnotParametrizationValues();
+
+    ASSERT_EQ(4, parametrization_values.size());
+
+    for (size_t i = 0; i < 2; i++)
+    {
+        const double knot_input_value = parametrization_values[i + 1];
```

Comment:
super nit: replace `[]` with `.at()` so we get an exception if the test_spline changes and we accidentally access memory out of bounds (will hopefully be more obvious in test logs)

Reply:
yep, changed